pkg/services/ngalert/notifier/locking_config.go

```diff
@@ -0,0 +1,80 @@
+package notifier
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	api "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+)
+
+var (
+	ErrBadAlertmanagerConfiguration = fmt.Errorf("bad Alertmanager configuration")
+)
+
+type AlertmanagerLockedConfig struct {
+	Config        *api.PostableUserConfig
+	ConfigHash    string
+	ConfigVersion string
+}
+
+type LockingConfigStore struct {
```

Comment:
Can you please explain what "Locking" stands for?

Reply:
I'll add a comment explaining it 👍 

For reference, this is borrowed from the provisioning `alertmanagerConfigStoreImpl`. We optimistically lock based on the hash of the config, hence "Locking". Technically for now this service could use the bare `GetLatestAlertmanagerConfiguration()` and deserialize the config string since it only reads the config. I'd be open to moving this to a later PR when we're actually writing back the config, WDYT?

Reply:
This particular struct does not implement locking, the underlying store does. It's just a convenience wrapper that handles serialization. So, I do not we should use term "locking". Perhaps "Structured" or "Typed"? 

It would be great if we could refactor instead of copying. Perhaps a quick refactor PR could do it. WDYT?

Reply:
we can move those two methods to store package

Reply:
I was hesitant to just add these to the `store` package because they're using api models, but yeah instead of worrying about names we can just do that.